pkg/server/home_workspaces.go

```diff
@@ -0,0 +1,661 @@
+			_, err := kubeClusterClient.Cluster(workspace).RbacV1().ClusterRoleBindings().Create(ctx, crb, metav1.CreateOptions{})
+			return err
+		},
+		createClusterWorkspace: func(ctx context.Context, workspace logicalcluster.Name, cw *tenancyv1alpha1.ClusterWorkspace) error {
+			_, err := kcpClusterClient.Cluster(workspace).TenancyV1alpha1().ClusterWorkspaces().Create(ctx, cw, metav1.CreateOptions{})
+			return err
+		},
+		getClusterWorkspace: func(ctx context.Context, workspace logicalcluster.Name, name string) (*tenancyv1alpha1.ClusterWorkspace, error) {
+			return kcpClusterClient.Cluster(workspace).TenancyV1alpha1().ClusterWorkspaces().Get(ctx, name, metav1.GetOptions{})
+		},
+	}
+}
+
+func addWorkspaceFullPathIndexer(clusterWorkspaceInformer cache.SharedIndexInformer) error {
+	return clusterWorkspaceInformer.AddIndexers(cache.Indexers{
+		workspaceFullPathIndexName: func(obj interface{}) ([]string, error) {
+			o, ok := obj.(*tenancyv1alpha1.ClusterWorkspace)
+			if !ok {
+				return []string{}, fmt.Errorf("obj is supposed to be a pointer to ClusterWorkspace, but is %T", obj)
+			}
+
+			cluster := logicalcluster.From(o)
+			return []string{cluster.Join(o.Name).String()}, nil
+		},
+	})
+}
+
+type searchFunc func(workspace logicalcluster.Name, name string) (bool, error)
+type mockableFunctions struct {
+	searchHomeWorkspaceRBACLocally func(homeWorkspace logicalcluster.Name, userName string) (found bool, err error)
+	createHomeRBAC                 func(ctx context.Context, userName string, homeWorkspace logicalcluster.Name) error
+	checkWorkspaceLocally          func(workspaceName logicalcluster.Name, isHome bool, userName string) (found, retryLater bool, retryAfterSeconds int, checkError error)
+	tryToCreate                    func(ctx context.Context, userName string, workspaceToCheck logicalcluster.Name, workspaceType tenancyv1alpha1.ClusterWorkspaceTypeName) (retryLater bool, retryAfterSeconds int, createError error)
+}
+
+type homeWorkspaceHandler struct {
+	apiHandler http.Handler
+
+	authz authorizer.Authorizer
+
+	createClusterRole        func(ctx context.Context, workspace logicalcluster.Name, cr *rbacv1.ClusterRole) error
+	createClusterRoleBinding func(ctx context.Context, workspace logicalcluster.Name, crb *rbacv1.ClusterRoleBinding) error
+
+	createClusterWorkspace func(ctx context.Context, workspace logicalcluster.Name, cw *tenancyv1alpha1.ClusterWorkspace) error
+	getClusterWorkspace    func(ctx context.Context, workspace logicalcluster.Name, name string) (*tenancyv1alpha1.ClusterWorkspace, error)
+
+	getLocalClusterWorkspace      func(fullName logicalcluster.Name) (*tenancyv1alpha1.ClusterWorkspace, error)
+	searchLocalClusterRole        searchFunc
+	searchLocalClusterRoleBinding searchFunc
+	synced                        func() bool
+
+	homePrefix               logicalcluster.Name
+	bucketLevels, bucketSize int
+	creationDelay            int
```

Comment:
time.Duration. And what is this?

Reply:
This is expected to be in seconds: the number of seconds we'd ask before a retry when the workspace has to be created or waited for readiness.
I renamed it to make it explicit it is a number of seconds in commit 1f37994d81146bb38fbecd97319d1a5a24eba461